File: src/workspace.mjs
import { execFileSync, spawn } from 'node:child_process';
import * as fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

export function git(args, cwd) {
  try {
    return execFileSync('git', ['-c', 'protocol.ext.allow=never', ...args], {
      cwd, encoding: 'utf8', windowsHide: true,
      stdio: ['inherit', 'pipe', 'pipe'], maxBuffer: 8 * 1024 * 1024,
    }).trim();
  } catch (error) {
    if (error.code === 'ENOENT') throw new Error('Git was not found. Install Git and check your PATH.');
    const detail = String(error.stderr || error.message).replace(/(https?:\/\/)[^\s/@]+@/g, '$1[redacted]@');
    throw new Error(`Git command failed: ${detail.trim()}`);
  }
}

function repositoryKind(value) {
  if (typeof value !== 'string' || !value.trim() || /[\r\n\0]/.test(value)) {
    throw new Error('A repository reference must be a non-empty URL or path.');
  }
  if (/^[A-Za-z][A-Za-z0-9+.-]*::/.test(value)) throw new Error('Git remote helpers are not allowed.');
  if (/^[A-Za-z][A-Za-z0-9+.-]*:\/\//.test(value)) {
    const url = new URL(value);
    if (!['https:', 'ssh:', 'file:'].includes(url.protocol)) throw new Error('Use HTTPS, SSH or a local Git path.');
    if (url.password || (url.protocol === 'https:' && url.username) || url.search || url.hash) {
      throw new Error('Do not put credentials, query parameters or fragments in repository URLs.');
    }
    return url.protocol === 'file:' ? 'file' : 'url';
  }
  if (/^[^\s/@:]+@[^\s/:]+:.+$/.test(value)) return 'scp';
  return 'local';
}

export function normalizeRepository(value, cwd = process.cwd()) {
  const kind = repositoryKind(value);
  if (kind === 'file') return fileURLToPath(value);
  return kind === 'local' ? path.resolve(cwd, value) : value;
}

export function resolveRepository(reference, specRepository) {
  const kind = repositoryKind(reference);
  const baseKind = repositoryKind(specRepository);
  const relative = reference.startsWith('./') || reference.startsWith('../');
  if (!relative) {
    if (['url', 'scp'].includes(baseKind) && ['local', 'file'].includes(kind)) {
      throw new Error('A remote spec cannot reference an absolute local repository path.');
    }
    return normalizeRepository(reference);
  }
  if (baseKind === 'url') return new URL(reference, specRepository.replace(/\/$/, '') + '/').href;
  if (baseKind === 'scp') {
    const colon = specRepository.indexOf(':');
    return specRepository.slice(0, colon + 1) + path.posix.normalize(path.posix.join(specRepository.slice(colon + 1), reference));
  }
  return path.resolve(normalizeRepository(specRepository), reference);
}

const samePath = (a, b) => {
  const normalize = value => process.platform === 'win32' ? path.resolve(value).toLowerCase() : path.resolve(value);
  return normalize(a) === normalize(b);
};

function portableName(value) {
  return typeof value === 'string' && /^[A-Za-z0-9][A-Za-z0-9._-]*$/.test(value)
    && !value.endsWith('.') && !/^(con|prn|aux|nul|com[1-9]|lpt[1-9])(?:\.|$)/i.test(value);
}

export function repositoryRoot(directory) {
  if (typeof directory !== 'string' || !directory) throw new Error('A repository directory is required.');
  const absolute = path.resolve(directory);
  const root = path.resolve(git(['rev-parse', '--show-toplevel'], absolute));
  if (!samePath(root, absolute)) throw new Error(`Use the repository root: ${root}`);
  return root;
}

export function originOrLocal(directory) {
  try { return normalizeRepository(git(['remote', 'get-url', 'origin'], directory), directory); }
  catch { return directory; }
}

export function normalizeNamespace(value, cwd = process.cwd()) {
  if (portableName(value)) return `git@github.com:${value}`;
  const normalized = normalizeRepository(value, cwd);
  return repositoryKind(normalized) === 'local' ? normalized : normalized.replace(/\/+$/, '');
}

export function namespaceFromRepository(repository) {
  return resolveRepository('../', normalizeRepository(repository));
}

export function resolveSpecSource(input, repositoryNamespace, cwd = process.cwd()) {
  if (portableName(input)) {
    if (!repositoryNamespace) throw new Error('Configure a repository namespace first: cspec config namespace <GitHub-owner-or-namespace-URL>');
    const base = normalizeNamespace(repositoryNamespace, cwd);
    return repositoryKind(base) === 'local' ? path.join(base, input) : base + '/' + input;
  }
  return normalizeRepository(input, cwd);
}

function repositoryDirectory(source) {
  const kind = repositoryKind(source);
  const value = kind === 'url' ? new URL(source).pathname : kind === 'scp' ? source.slice(source.indexOf(':') + 1) : source;
  const name = (kind === 'local' ? path.basename(value) : path.posix.basename(value.replace(/\/$/, ''))).replace(/\.git$/, '');
  if (!portableName(name)) throw new Error('This repository needs an explicit destination directory with a portable name.');
  return name;
}

function objectKeys(value, expected, label) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) throw new Error(`${label} must be an object.`);
  if (Object.keys(value).sort().join(',') !== [...expected].sort().join(',')) {
    throw new Error(`${label} requires exactly: ${expected.join(', ')}.`);
  }
}

async function readJson(filename) {
  const info = await fs.lstat(filename);
  if (!info.isFile() || info.size > 128 * 1024) throw new Error(`Invalid configuration file: ${filename}`);
  return JSON.parse(await fs.readFile(filename, 'utf8'));
}

export async function readProject(specDirectory) {
  const manifest = await readJson(path.join(specDirectory, 'project.json'));
  const lock = await readJson(path.join(specDirectory, 'guidelines.lock.json'));
  objectKeys(manifest, ['schemaVersion', 'name', 'code', 'guidelines', 'profile'], 'project.json');
  objectKeys(manifest.code, ['repository'], 'code');
  objectKeys(manifest.guidelines, ['repository', 'ref'], 'guidelines');
  objectKeys(lock, ['schemaVersion', 'ref', 'commit'], 'guidelines.lock.json');
  if (manifest.schemaVersion !== 1 || lock.schemaVersion !== 1) throw new Error('Only schemaVersion 1 is supported.');
  if (!portableName(manifest.name) || !/^[A-Za-z]/.test(manifest.name)) throw new Error('Invalid or non-portable project name.');
  if (typeof manifest.profile !== 'string' || !/^[A-Za-z][A-Za-z0-9_-]*$/.test(manifest.profile)) throw new Error('Invalid profile.');
  repositoryKind(manifest.code.repository);
  repositoryKind(manifest.guidelines.repository);
  if (typeof manifest.guidelines.ref !== 'string' || !/^[A-Za-z0-9][A-Za-z0-9._/-]*$/.test(manifest.guidelines.ref)) throw new Error('Invalid guidelines ref.');
  if (lock.ref !== manifest.guidelines.ref || typeof lock.commit !== 'string' || !/^([0-9a-f]{40}|[0-9a-f]{64})$/.test(lock.commit)) {
    throw new Error('The lock must contain the same ref and a full exact commit ID.');
  }
  const specInfo = await fs.lstat(path.join(specDirectory, 'spec'));
  if (!specInfo.isDirectory()) throw new Error('The spec/ directory is missing or is not a regular directory.');
  return { manifest, lock };
}

async function exists(filename) {
  try { await fs.lstat(filename); return true; }
  catch (error) { if (error.code === 'ENOENT') return false; throw error; }
}

async function prepareParent(directory) {
  const target = path.resolve(directory);
  let parent = target;
  while (!(await exists(parent))) parent = path.dirname(parent);
  let gitRoot;
  try { gitRoot = git(['rev-parse', '--show-toplevel'], parent); } catch { /* Standalone clones have no enclosing Git repository. */ }
  if (gitRoot) throw new Error('Create the project outside existing Git repositories.');
  await fs.mkdir(target, { recursive: true });
  return target;
}

async function createTarget(directory) {
  const target = path.resolve(directory);
  if (await exists(target)) throw new Error(`Destination already exists; nothing overwritten: ${target}`);
  await prepareParent(path.dirname(target));
  await fs.mkdir(target);
  return target;
}

function codePath(spec, manifest) {
  const code = path.join(path.dirname(spec), manifest.name);
  if (samePath(code, spec)) throw new Error('The spec directory and the sibling code directory must have different names.');
  return code;
}

function projectPaths(spec, manifest, source = originOrLocal(spec)) {
  const projectRoot = path.dirname(spec);
  const code = codePath(spec, manifest);
  const guidelinesRepository = resolveRepository(manifest.guidelines.repository, source);
  const editableGuidelines = path.join(projectRoot, repositoryDirectory(guidelinesRepository));
  const names = [spec, code, editableGuidelines].map(directory => path.basename(directory).toLowerCase());
  if (new Set(names).size !== names.length) throw new Error('Spec, code and guidelines need distinct directory names inside the project root.');
  return { projectRoot, code, editableGuidelines, guidelinesRepository };
}

function repositoryIdentity(reference) {
  const source = normalizeRepository(reference);
  const kind = repositoryKind(source);
  if (kind === 'url') {
    const url = new URL(source);
    const port = url.port && !['22', '443'].includes(url.port) ? ':' + url.port : '';
    return url.hostname.toLowerCase() + port + url.pathname.replace(/\/$/, '').replace(/\.git$/, '');
  }
  if (kind === 'scp') {
    const hostStart = source.indexOf('@') + 1;
    const colon = source.indexOf(':', hostStart);
    return source.slice(hostStart, colon).toLowerCase() + '/' + source.slice(colon + 1).replace(/\/$/, '').replace(/\.git$/, '');
  }
  return process.platform === 'win32' ? path.resolve(source).toLowerCase() : path.resolve(source);
}

async function localDirectory(spec) {
  if (git(['ls-files', '--', '.local'], spec)) throw new Error('The spec tracks .local/ files. Untrack those files before generating local context.');
  const local = path.join(spec, '.local');
  if (await exists(local)) {
    if (!(await fs.lstat(local)).isDirectory()) throw new Error('The spec .local/ path must be a regular directory, not a link.');
  } else {
    await fs.mkdir(local);
  }
  try { git(['check-ignore', '--quiet', '--', '.local/cspec-probe'], spec); }
  catch {
    const exclude = path.resolve(spec, git(['rev-parse', '--git-path', 'info/exclude'], spec));
    await fs.mkdir(path.dirname(exclude), { recursive: true });
    await fs.appendFile(exclude, '\n# CretSpec generated local files\n/.local/\n');
  }
  return local;
}

async function guidelinesSnapshot(spec, manifest, lock, progress) {
  const { editableGuidelines, guidelinesRepository } = projectPaths(spec, manifest);
  if (!(await exists(editableGuidelines))) {
    progress('Fetching project guidelines');
    git(['clone', '--no-recurse-submodules', '--no-hardlinks', '--', guidelinesRepository, editableGuidelines]);
  }
  if (!(await fs.lstat(editableGuidelines)).isDirectory()) throw new Error('The project guidelines must be a regular directory.');
  repositoryRoot(editableGuidelines);
  if (repositoryIdentity(originOrLocal(editableGuidelines)) !== repositoryIdentity(guidelinesRepository)) {
    throw new Error('The project guidelines repository does not match guidelines.repository in the spec.');
  }
  try {
    if (git(['cat-file', '-t', lock.commit], editableGuidelines) !== 'commit') throw new Error('Not a commit.');
  } catch {
    throw new Error(`The project guidelines repository does not contain ${lock.commit}.\nRun git fetch --tags in ${editableGuidelines}.`);
  }
  if (git(['rev-parse', '--verify', `${lock.ref}^{commit}`], editableGuidelines) !== lock.commit) {
    throw new Error('The guidelines ref does not match the lock. No version was adopted.');
  }
  const local = await localDirectory(spec);
  const cache = path.join(local, 'guidelines');
  if (await exists(cache)) {
    if (!(await fs.lstat(cache)).isDirectory()) throw new Error('The guidelines cache must be a regular directory.');
  } else {
    await fs.mkdir(cache);
  }
  const activeGuidelines = path.join(cache, lock.commit);
  if (!(await exists(activeGuidelines))) {
    progress('Fetching pinned guidelines');
    git(['clone', '--no-recurse-submodules', '--no-hardlinks', '--', editableGuidelines, activeGuidelines]);
    const referencedCommit = git(['rev-parse', '--verify', `${lock.ref}^{commit}`], activeGuidelines);
    if (referencedCommit !== lock.commit) throw new Error('The guidelines ref does not match the lock. No version was adopted.');
    git(['checkout', '--detach', lock.commit], activeGuidelines);
  }
  if (!(await fs.lstat(activeGuidelines)).isDirectory()) throw new Error('The guidelines snapshot must be a regular directory.');
  repositoryRoot(activeGuidelines);
  if (git(['rev-parse', 'HEAD'], activeGuidelines) !== lock.commit
    || git(['rev-parse', '--verify', `${lock.ref}^{commit}`], activeGuidelines) !== lock.commit
    || git(['status', '--porcelain', '--untracked-files=all'], activeGuidelines)) {
    throw new Error('The cached guidelines differ from the lock or contain local changes. Preserve any work and remove that snapshot before retrying.');
  }
  const profile = await fs.lstat(path.join(activeGuidelines, 'profiles', `${manifest.profile}.md`));
  if (!profile.isFile()) throw new Error('The selected profile must be a regular file.');
  return { editableGuidelines, activeGuidelines };
}

export async function findSpec(start = process.cwd()) {
  let current = path.resolve(start);
  if ((await fs.stat(current)).isFile()) current = path.dirname(current);
  while (true) {
    if (await exists(path.join(current, 'project.json')) && await exists(path.join(current, 'guidelines.lock.json'))) {
      const spec = repositoryRoot(current);
      await readProject(spec);
      return spec;
    }
    let enclosingRepository;
    try { enclosingRepository = path.resolve(git(['rev-parse', '--show-toplevel'], current)); } catch { /* An ordinary project container is not a Git repository. */ }
    if (enclosingRepository && !samePath(current, enclosingRepository)) {
      current = enclosingRepository;
      continue;
    }
    if (!enclosingRepository) {
      const candidates = [];
      for (const entry of await fs.readdir(current, { withFileTypes: true })) {
        if (!entry.isDirectory() || entry.name.startsWith('.')) continue;
        const directory = path.join(current, entry.name);
        if (await exists(path.join(directory, 'project.json')) && await exists(path.join(directory, 'guidelines.lock.json'))) candidates.push(directory);
      }
      if (candidates.length > 1) throw new Error('Multiple project specs found. Pass the intended spec directory explicitly.');
      if (candidates.length === 1) {
        const spec = repositoryRoot(candidates[0]);
        await readProject(spec);
        return spec;
      }
    }
    const parent = path.dirname(current);
    if (parent === current) throw new Error('No project spec found. Run this command inside a project root, code or spec repository, or pass a project directory.');
    current = parent;
  }
}

export async function projectInfo(start, { progress = () => {} } = {}) {
  const spec = await findSpec(start);
  const { manifest, lock } = await readProject(spec);
  const { projectRoot, code: codeDirectory } = projectPaths(spec, manifest);
  const code = repositoryRoot(codeDirectory);
  const expectedSource = resolveRepository(manifest.code.repository, originOrLocal(spec));
  if (repositoryIdentity(originOrLocal(code)) !== repositoryIdentity(expectedSource)) {
    throw new Error(`The sibling code repository does not match code.repository in the spec: ${code}\nExpected source: ${expectedSource}\nUse the correct clone; the existing repository has not been changed.`);
  }
  const guidelines = await guidelinesSnapshot(spec, manifest, lock, progress);
  return { projectRoot, spec, code, ...guidelines, manifest, lock };
}

export async function cloneProject(input, directory, { repositoryNamespace, cwd = process.cwd(), progress = () => {} } = {}) {
  const source = resolveSpecSource(input, repositoryNamespace, cwd);
  const explicitRoot = directory ? path.resolve(cwd, directory) : null;
  if (explicitRoot && await exists(explicitRoot)) throw new Error(`Destination already exists; nothing overwritten: ${explicitRoot}`);
  const parent = await prepareParent(explicitRoot ? path.dirname(explicitRoot) : cwd);
  const staging = await fs.mkdtemp(path.join(parent, '.cspec-clone-'));
  let projectRoot;
  let spec;
  try {
    progress('Fetching project spec');
    git(['clone', '--no-recurse-submodules', '--no-hardlinks', '--', source, staging]);
    const { manifest } = await readProject(staging);
    const target = explicitRoot || path.join(parent, manifest.name);
    const targetSpec = path.join(target, repositoryDirectory(source));
    const { code } = projectPaths(targetSpec, manifest, source);
    projectRoot = await createTarget(target);
    await fs.rename(staging, targetSpec);
    spec = targetSpec;
    progress('Fetching code');
    git(['clone', '--no-recurse-submodules', '--no-hardlinks', '--', resolveRepository(manifest.code.repository, source), code]);
    return await projectInfo(spec, { progress });
  } catch (error) {
    const preserved = [projectRoot, !spec ? staging : null].filter(Boolean).join(', ');
    throw new Error(`${error.message}\nNew directories preserved for inspection: ${preserved}\nExisting repositories have not been modified.`);
  }
}

export async function attachProject(codeDirectory, specDirectory, { progress = () => {} } = {}) {
  const spec = repositoryRoot(specDirectory);
  const code = repositoryRoot(codeDirectory);
  const { manifest } = await readProject(spec);
  if (!samePath(code, codePath(spec, manifest))) {
    throw new Error(`The spec expects its code in the sibling directory ${codePath(spec, manifest)}. No separate project bindings are stored.`);
  }
  return projectInfo(spec, { progress });
}

export async function workspacePath(start, options = {}) {
  const info = await projectInfo(start, options);
  const local = await localDirectory(info.spec);
  const workspaceFile = path.join(local, 'project.code-workspace');
  let existing = {};
  if (await exists(workspaceFile)) existing = await readJson(workspaceFile);
  const relative = directory => path.relative(local, directory).split(path.sep).join('/');
  await fs.writeFile(workspaceFile, JSON.stringify({
    ...existing,
    folders: [
      { name: path.basename(info.code), path: relative(info.code) },
      { name: path.basename(info.spec), path: relative(info.spec) },
      { name: path.basename(info.editableGuidelines), path: relative(info.editableGuidelines) },
    ],
  }, null, 2) + '\n');
  return workspaceFile;
}

export async function openPath(filename) {
  const absolute = path.resolve(filename);
  await fs.access(absolute);
  const command = process.platform === 'win32' ? 'explorer.exe' : process.platform === 'darwin' ? 'open' : 'xdg-open';
  await new Promise((resolve, reject) => {
    const child = spawn(command, [absolute], { shell: false, windowsHide: true, stdio: 'ignore', detached: true });
    child.once('error', () => reject(new Error(`Open the path manually; ${command} could not be started: ${absolute}`)));
    child.once('spawn', () => { child.unref(); resolve(); });
  });
}

File: src/workspace.rs
use crate::{
    files, git,
    manifest::{self, GuidelinesMode, Lock, Manifest},
    repository,
};
use anyhow::{Context, Result, bail};
use serde::Serialize;
use serde_json::{Value, json};
use std::{
    fs,
    io::Write,
    path::{Path, PathBuf},
    process::Stdio,
};

pub type Progress<'a> = &'a dyn Fn(&str);

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectInfo {
    pub project_root: PathBuf,
    pub spec: PathBuf,
    pub code: PathBuf,
    pub editable_guidelines: PathBuf,
    pub active_guidelines: PathBuf,
    pub manifest: Manifest,
    pub lock: Option<Lock>,
    pub guidelines: GuidelinesState,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GuidelinesState {
    pub mode: GuidelinesMode,
    pub commit: String,
    pub branch: Option<String>,
    pub dirty: bool,
    pub upstream: Option<String>,
    pub ahead: Option<u64>,
    pub behind: Option<u64>,
}

pub fn guidelines_state(active: &Path, mode: GuidelinesMode) -> Result<GuidelinesState> {
    let commit = git::run(active, ["rev-parse", "HEAD"])?;
    let branch = git::run(active, ["symbolic-ref", "--quiet", "--short", "HEAD"]).ok();
    let dirty = !git::run(active, ["status", "--porcelain", "--untracked-files=all"])?.is_empty();
    let upstream = git::run(
        active,
        [
            "rev-parse",
            "--abbrev-ref",
            "--symbolic-full-name",
            "@{upstream}",
        ],
    )
    .ok();
    let counts = upstream.as_ref().and_then(|_| {
        git::run(
            active,
            ["rev-list", "--left-right", "--count", "HEAD...@{upstream}"],
        )
        .ok()
    });
    let counts = counts
        .as_deref()
        .and_then(|s| s.split_once(char::is_whitespace));
    let (ahead, behind) = counts
        .map(|(a, b)| (a.trim().parse().ok(), b.trim().parse().ok()))
        .unwrap_or((None, None));
    Ok(GuidelinesState {
        mode,
        commit,
        branch,
        dirty,
        upstream,
        ahead,
        behind,
    })
}

pub fn validate_profile(active: &Path, profile: &str) -> Result<()> {
    files::directory(&active.join("profiles"))?;
    let path = active.join("profiles").join(format!("{profile}.md"));
    if !fs::symlink_metadata(&path)
        .with_context(|| format!("Selected profile does not exist: {}", path.display()))?
        .file_type()
        .is_file()
    {
        bail!("The selected profile must be a regular file.");
    }
    Ok(())
}

pub fn validate_branch(directory: &Path, branch: &str) -> Result<()> {
    git::run(
        directory,
        ["check-ref-format", &format!("refs/heads/{branch}")],
    )
    .context("Working-tree guidelines need a valid branch name")?;
    let local = format!("refs/heads/{branch}");
    let remote = format!("refs/remotes/origin/{branch}");
    if git::run(directory, ["show-ref", "--verify", &local]).is_err()
        && git::run(directory, ["show-ref", "--verify", &remote]).is_err()
    {
        bail!(
            "Working-tree guidelines need an available branch, not a tag or commit: {branch}. Fetch the branch with Git or restore a valid branch definition before pinning a version with cspec guidelines update <ref>."
        );
    }
    Ok(())
}

pub struct Paths {
    pub root: PathBuf,
    pub code: PathBuf,
    pub guidelines: PathBuf,
    pub guidelines_source: String,
}

pub fn paths(spec: &Path, manifest: &Manifest, source: &str) -> Result<Paths> {
    let root = spec
        .parent()
        .context("Spec has no parent directory")?
        .to_owned();
    let code = root.join(&manifest.name);
    let guidelines_source = repository::resolve(&manifest.guidelines.repository, source)?;
    let guidelines = root.join(repository::directory_name(&guidelines_source)?);
    let mut names = [spec, code.as_path(), guidelines.as_path()]
        .iter()
        .map(|p| {
            p.file_name()
                .unwrap_or_default()
                .to_string_lossy()
                .to_lowercase()
        })
        .collect::<Vec<_>>();
    names.sort();
    names.dedup();
    if names.len() != 3 {
        bail!("Spec, code and guidelines need distinct directory names inside the project root.");
    }
    Ok(Paths {
        root,
        code,
        guidelines,
        guidelines_source,
    })
}

pub fn local_directory(spec: &Path, create: bool) -> Result<PathBuf> {
    if !git::run(spec, ["ls-files", "--", ".local"])?.is_empty() {
        bail!(
            "The spec tracks .local/ files. Untrack those files before generating local context."
        );
    }
    let local = spec.join(".local");
    if files::exists(&local)? {
        files::directory(&local)?;
    } else if create {
        fs::create_dir(&local)?;
    } else {
        bail!("Local project context is missing. Run cspec project info to prepare it.");
    }
    if create
        && git::run(
            spec,
            ["check-ignore", "--quiet", "--", ".local/cspec-probe"],
        )
        .is_err()
    {
        let exclude = files::absolute(
            spec.join(git::run(spec, ["rev-parse", "--git-path", "info/exclude"])?),
        )?;
        fs::create_dir_all(exclude.parent().context("Git exclude path has no parent")?)?;
        let mut file = fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(exclude)?;
        file.write_all(b"\n# Generated local files\n/.local/\n")?;
    }
    Ok(local)
}

pub fn verify_repository(directory: &Path, expected: &str, label: &str) -> Result<()> {
    git::root(directory)?;
    if repository::identity(&repository::origin(directory)?)? != repository::identity(expected)? {
        bail!(
            "The {label} repository does not match its source in the spec: {}\nExpected source: {expected}\nUse the correct clone; the existing repository has not been changed.",
            directory.display()
        );
    }
    Ok(())
}

pub fn verify_revision(directory: &Path, lock: &Lock) -> Result<()> {
    if git::run(directory, ["cat-file", "-t", &lock.commit])
        .ok()
        .as_deref()
        != Some("commit")
    {
        bail!(
            "The project guidelines repository does not contain {}.\nRun git fetch --tags in {}.",
            lock.commit,
            directory.display()
        );
    }
    let revision = git::run(
        directory,
        [
            "rev-parse",
            "--verify",
            &format!("{}^{{commit}}", lock.reference),
        ],
    )?;
    if revision != lock.commit {
        bail!("The guidelines ref does not match the lock. No version was adopted.");
    }
    Ok(())
}

pub fn snapshot(
    spec: &Path,
    manifest: &Manifest,
    lock: Option<&Lock>,
    create: bool,
    progress: Progress<'_>,
) -> Result<(PathBuf, PathBuf)> {
    let layout = paths(spec, manifest, &repository::origin(spec)?)?;
    let new_clone = !files::exists(&layout.guidelines)?;
    if new_clone {
        if !create {
            bail!(
                "The editable guidelines clone is missing. Run cspec project info to prepare it."
            );
        }
        progress("Fetching project guidelines");
        git::clone(&layout.guidelines_source, &layout.guidelines)?;
    }
    verify_repository(&layout.guidelines, &layout.guidelines_source, "guidelines")?;
    if manifest.guidelines_mode(lock) == GuidelinesMode::WorkingTree {
        validate_branch(&layout.guidelines, &manifest.guidelines.reference)?;
        if new_clone {
            git::run(
                &layout.guidelines,
                ["switch", &manifest.guidelines.reference],
            )?;
        }
        validate_profile(&layout.guidelines, &manifest.profile)?;
        if create {
            local_directory(spec, true)?;
        }
        return Ok((layout.guidelines.clone(), layout.guidelines));
    }
    let lock = lock.context("Pinned guidelines require a lock")?;
    verify_revision(&layout.guidelines, lock)?;
    let cache = local_directory(spec, create)?.join("guidelines");
    if files::exists(&cache)? {
        files::directory(&cache)?;
    } else if create {
        fs::create_dir(&cache)?;
    }
    let active = cache.join(&lock.commit);
    if !files::exists(&active)? {
        if !create {
            bail!(
                "The pinned guidelines snapshot is missing. Run cspec project info to prepare it."
            );
        }
        progress("Fetching pinned guidelines");
        git::clone(&layout.guidelines.to_string_lossy(), &active)?;
        verify_revision(&active, lock)?;
        git::run(&active, ["checkout", "--detach", &lock.commit])?;
    }
    git::root(&active)?;
    let head = git::run(&active, ["rev-parse", "HEAD"])?;
    let dirty = git::run(&active, ["status", "--porcelain", "--untracked-files=all"])?;
    let attached = git::run(&active, ["symbolic-ref", "--quiet", "HEAD"]).is_ok();
    if head != lock.commit
        || !dirty.is_empty()
        || attached
        || verify_revision(&active, lock).is_err()
    {
        bail!(
            "The cached guidelines differ from the lock or contain local changes. Preserve any work and remove that snapshot before retrying."
        );
    }
    validate_profile(&active, &manifest.profile)?;
    Ok((layout.guidelines, active))
}

fn is_spec(path: &Path) -> Result<bool> {
    Ok(files::exists(&path.join("project.json"))? && files::exists(&path.join("spec"))?)
}

pub fn find_spec(start: &Path) -> Result<PathBuf> {
    let spec = discover_spec(start)?;
    manifest::read(&spec)?;
    Ok(spec)
}

pub fn discover_spec(start: &Path) -> Result<PathBuf> {
    let mut current = files::absolute(start)?;
    if fs::metadata(&current)?.is_file() {
        current.pop();
    }
    loop {
        if is_spec(&current)? {
            let spec = git::root(&current)?;
            return Ok(spec);
        }
        let enclosing = git::enclosing(&current);
        if let Some(root) = &enclosing {
            if !files::same(root, &current) {
                current = root.clone();
                continue;
            }
        } else {
            let mut candidates = Vec::new();
            for entry in fs::read_dir(&current)? {
                let entry = entry?;
                if !entry.file_type()?.is_dir()
                    || entry.file_name().to_string_lossy().starts_with('.')
                {
                    continue;
                }
                if is_spec(&entry.path())? {
                    candidates.push(entry.path());
                }
            }
            if candidates.len() > 1 {
                bail!("Multiple project specs found. Pass the intended spec directory explicitly.");
            }
            if let Some(candidate) = candidates.first() {
                let spec = git::root(candidate)?;
                return Ok(spec);
            }
        }
        if !current.pop() {
            bail!(
                "No project spec found. Run this command inside a project root, code or spec repository, or pass a project directory."
            );
        }
    }
}

pub fn info(start: &Path, create: bool, progress: Progress<'_>) -> Result<ProjectInfo> {
    let spec = find_spec(start)?;
    let (manifest, lock) = manifest::read(&spec)?;
    let source = repository::origin(&spec)?;
    let layout = paths(&spec, &manifest, &source)?;
    verify_repository(
        &layout.code,
        &repository::resolve(&manifest.code.repository, &source)?,
        "code",
    )?;
    let (editable_guidelines, active_guidelines) =
        snapshot(&spec, &manifest, lock.as_ref(), create, progress)?;
    let guidelines = guidelines_state(&active_guidelines, manifest.guidelines_mode(lock.as_ref()))?;
    Ok(ProjectInfo {
        project_root: layout.root,
        spec,
        code: layout.code,
        editable_guidelines,
        active_guidelines,
        manifest,
        lock,
        guidelines,
    })
}

fn prepare_parent(directory: &Path) -> Result<PathBuf> {
    let target = files::absolute(directory)?;
    let mut parent = target.clone();
    while !files::exists(&parent)? {
        if !parent.pop() {
            bail!("Cannot locate an existing parent directory.");
        }
    }
    if git::enclosing(&parent).is_some() {
        bail!("Create the project outside existing Git repositories.");
    }
    fs::create_dir_all(&target)?;
    Ok(target)
}

pub fn clone_project(
    input: &str,
    destination: Option<&Path>,
    namespace: Option<&str>,
    cwd: &Path,
    progress: Progress<'_>,
) -> Result<ProjectInfo> {
    let source = repository::spec_source(input, namespace, cwd)?;
    let explicit = destination
        .map(|d| files::absolute(cwd.join(d)))
        .transpose()?;
    if let Some(target) = &explicit
        && files::exists(target)?
    {
        bail!(
            "Destination already exists; nothing overwritten: {}",
            target.display()
        );
    }
    let parent = prepare_parent(explicit.as_deref().and_then(Path::parent).unwrap_or(cwd))?;
    let staging = tempfile::Builder::new()
        .prefix(".cspec-clone-")
        .tempdir_in(&parent)?
        .keep();
    let mut preserved = staging.clone();
    let result = (|| -> Result<ProjectInfo> {
        progress("Fetching project spec");
        git::clone(&source, &staging)?;
        let (manifest, _) = manifest::read(&staging)?;
        let root = explicit.unwrap_or_else(|| parent.join(&manifest.name));
        if files::exists(&root)? {
            bail!(
                "Destination already exists; nothing overwritten: {}",
                root.display()
            );
        }
        let spec = root.join(repository::directory_name(&source)?);
        let layout = paths(&spec, &manifest, &source)?;
        fs::create_dir(&root)?;
        fs::rename(&staging, &spec).with_context(|| {
            format!(
                "Could not move the cloned spec into {}. Staging remains at {}",
                spec.display(),
                staging.display()
            )
        })?;
        preserved = root;
        progress("Fetching code");
        git::clone(
            &repository::resolve(&manifest.code.repository, &source)?,
            &layout.code,
        )?;
        let project = info(&spec, true, progress)?;
        crate::agents::sync(&project)?;
        Ok(project)
    })();
    result.with_context(|| format!("Project preparation failed. New directories preserved for inspection: {}\nExisting repositories have not been modified",preserved.display()))
}

pub fn attach(code: &Path, spec: &Path, progress: Progress<'_>) -> Result<ProjectInfo> {
    let code = git::root(&files::absolute(code)?)?;
    let spec = git::root(&files::absolute(spec)?)?;
    let (manifest, _) = manifest::read(&spec)?;
    let expected = spec
        .parent()
        .context("Spec has no parent")?
        .join(&manifest.name);
    if !files::same(&code, &expected) {
        bail!(
            "The spec expects its code in the sibling directory {}. No separate project bindings are stored.",
            expected.display()
        );
    }
    let project = info(&spec, true, progress)?;
    crate::agents::sync(&project)?;
    Ok(project)
}

pub fn editor_workspace(start: &Path) -> Result<PathBuf> {
    let info = info(start, true, &|_| {})?;
    crate::agents::sync(&info)?;
    let file = local_directory(&info.spec, true)?.join("project.code-workspace");
    let mut value: Value = if files::exists(&file)? {
        files::read_json(&file)?
    } else {
        json!({})
    };
    let settings = value
        .as_object_mut()
        .context("Editor workspace must be a JSON object")?;
    let mut folders = Vec::new();
    for folder in [&info.code, &info.spec, &info.editable_guidelines] {
        let name = folder
            .file_name()
            .context("Project folder has no name")?
            .to_string_lossy();
        let relative = if folder == &info.spec {
            "..".into()
        } else {
            format!("../../{name}")
        };
        folders.push(json!({"name":name,"path":relative}));
    }
    settings.insert("folders".into(), json!(folders));
    files::write_json(&file, &value)?;
    Ok(file)
}

pub fn open_path(path: &Path) -> Result<()> {
    let path = files::absolute(path)?;
    fs::metadata(&path)?;
    let opener = if cfg!(windows) {
        "explorer.exe"
    } else if cfg!(target_os = "macos") {
        "open"
    } else {
        "xdg-open"
    };
    git::command(opener)
        .arg(&path)
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
        .with_context(|| {
            format!(
                "Open the path manually; {opener} could not be started: {}",
                path.display()
            )
        })?;
    Ok(())
}
